src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php

```diff
@@ -86,77 +94,92 @@ public function close(): bool
     {
         $ttl = ($this->ttl instanceof \Closure ? ($this->ttl)() : $this->ttl) ?? \ini_get('session.gc_maxlifetime');
-        $expiry = new UTCDateTime((time() + (int) $ttl) * 1000);
+        $expiry = $this->getUTCDateTime($ttl);
 
         $fields = [
-            $this->options['time_field'] => new UTCDateTime(),
+            $this->options['time_field'] => $this->getUTCDateTime(),
             $this->options['expiry_field'] => $expiry,
             $this->options['data_field'] => new Binary($data, Binary::TYPE_OLD_BINARY),
         ];
 
-        $this->getCollection()->updateOne(
+        $write = new BulkWrite();
+        $write->update(
             [$this->options['id_field'] => $sessionId],
             ['$set' => $fields],
             ['upsert' => true]
         );
 
+        $this->manager->executeBulkWrite($this->namespace, $write);
+
         return true;
     }
 
     public function updateTimestamp(#[\SensitiveParameter] string $sessionId, string $data): bool
     {
         $ttl = ($this->ttl instanceof \Closure ? ($this->ttl)() : $this->ttl) ?? \ini_get('session.gc_maxlifetime');
-        $expiry = new UTCDateTime((time() + (int) $ttl) * 1000);
+        $expiry = $this->getUTCDateTime($ttl);
 
-        $this->getCollection()->updateOne(
+        $write = new BulkWrite();
+        $write->update(
             [$this->options['id_field'] => $sessionId],
             ['$set' => [
-                $this->options['time_field'] => new UTCDateTime(),
+                $this->options['time_field'] => $this->getUTCDateTime(),
                 $this->options['expiry_field'] => $expiry,
-            ]]
+            ]],
+            ['limit' => 1],
         );
 
+        $this->manager->executeBulkWrite($this->namespace, $write);
+
         return true;
     }
 
     protected function doRead(#[\SensitiveParameter] string $sessionId): string
     {
-        $dbData = $this->getCollection()->findOne([
+        $dbData = $this->manager->executeQuery($this->namespace, new Query([
             $this->options['id_field'] => $sessionId,
-            $this->options['expiry_field'] => ['$gte' => new UTCDateTime()],
-        ]);
+            $this->options['expiry_field'] => ['$gte' => $this->getUTCDateTime()],
+        ]));
+        $dbData->setTypeMap(['root' => 'bson']);
+        $dbData->rewind();
+        $data = $dbData->current();
 
-        if (null === $dbData) {
+        if (!$data) {
             return '';
         }
 
-        return $dbData[$this->options['data_field']]->getData();
+        return $data->get($this->options['data_field']);
     }
 
-    private function getCollection(): Collection
+    private function getUTCDateTime(int $additionalSeconds = 0): UTCDateTime
     {
-        return $this->collection ??= $this->mongo->selectCollection($this->options['database'], $this->options['collection']);
-    }
+        $timestamp = (new \DateTimeImmutable('now', new \DateTimeZone('UTC')));
```

Comment:
This is time sensitive. What do you think of using the clock here, if available. That would help testing expiration.

Reply:
I wouldn't use the clock component to keep deps simple.

Reply:
Ok 👍🏻